provisioning/baremetal_pod.go

```diff
@@ -798,13 +782,29 @@ func EnsureMetal3Deployment(info *ProvisioningInfo) (updated bool, err error) {
 		info.EventRecorder, metal3Deployment, expectedGeneration)
 	if err != nil {
 		err = fmt.Errorf("unable to apply Metal3 deployment: %w", err)
-		return
+		// Check if ApplyDeployment failed because the existing Pod had an outdated
+		// Pod Selector.
+		selector, get_err := getMetal3DeploymentSelector(info.Client.AppsV1(), info.Namespace)
+		if get_err != nil || selector == metal3Deployment.Spec.Selector {
+			return
+		}
+		// This is an older deployment with the incorrect Pod Selector.
+		// Delete deployment and re-create
+		_ = DeleteMetal3Deployment(info)
+		_, create_err := info.Client.AppsV1().Deployments(info.Namespace).Create(
+			context.Background(),
+			metal3Deployment,
+			metav1.CreateOptions{})
```

Comment:
I personally wouldn't try to do a Create here. Just requeue and on the next reconcile ApplyDeployment will create it again. It's risky to try to create the Deployment in two different ways in two different places.

Reply:
Agreed. Fixed.